storage-transcribe-audio/functions/src/transcribe-audio.ts

```diff
@@ -0,0 +1,212 @@
+import { SpeechClient } from "@google-cloud/speech";
+import { google } from "@google-cloud/speech/build/protos/protos";
+import { Bucket } from "@google-cloud/storage";
+import * as ffmpeg from "fluent-ffmpeg";
+
+import {
+  TranscodeAudioResult,
+  WarningType,
+  FailureType,
+  TranscribeAudioResult,
+} from "./types";
+import * as logs from "./logs";
+import config from "./config";
+import { probePromise, isTaggedStringArray, separateByTags } from "./util";
+
+const encoding = "LINEAR16";
+const TRANSCODE_TARGET_FILE_EXTENSION = ".wav";
+
+export async function transcribeAndUpload({
+  client,
+  file: { bucket, name },
+  sampleRateHertz,
+  audioChannelCount,
+}: {
+  client: SpeechClient;
+  file: { bucket: Bucket; name: string };
+  sampleRateHertz: number;
+  audioChannelCount: number;
+}): Promise<TranscribeAudioResult> {
+  const warnings: WarningType[] = [];
+  const request: google.cloud.speech.v1.ILongRunningRecognizeRequest = {
+    config: {
+      encoding,
+      sampleRateHertz,
+      languageCode: config.languageCode,
+      model: config.model,
+      audioChannelCount,
+    },
+
+    audio: {
+      uri: `gs://${bucket.name}/${name}`,
+    },
+
+    outputConfig: {
+      gcsUri: `gs://${bucket.name}/${name}_transcription.txt`,
+    },
+  };
+
+  const response = await transcribe(client, request);
+
+  if (response.outputError) {
+    return {
+      state: "failure",
+      warnings,
+      type: FailureType.TRANSCRIPTION_UPLOAD_FAILED,
+      details: {
+        outputUri: response.outputConfig?.gcsUri,
+        outputError: response.outputError,
+      },
+    };
+  }
+
+  logs.receivedLongRunningRecognizeResponse(response);
+  const taggedTranscription = response.results?.map(
+    (result) =>
+      [result?.channelTag, result?.alternatives?.[0].transcript] as const
+  );
+
+  if (!isTaggedStringArray(taggedTranscription)) {
+    return {
+      state: "failure",
+      warnings,
+      type: FailureType.NULL_TRANSCRIPTION,
+    };
+  }
+
+  // I simplify the transcription compared to the one that's usually given
```

Comment:
Is this a TODO/will be removed before checking in the final version?

Reply:
Mainly a question for reviewers 😅

Reply:
I.e. I was asking the reviewers a question within the comment. What do you think of the question in the comment? We can also discuss it in 1:1 chat.

Reply:
I actually don't think I understand the problem/solution here. Can you explain more in depth?

Reply:
Yes.

I have two options: The extension can either (a) make the API call, get the transcription from it, and upload it itself, or (b) include an output path in the API call, so that the speech-to-text API uploads the file. In discussion, we agreed that going with (b) would likely be more reliable.

However, the output of the API looks something like this:
```json
{
  "results": [
    {
      "channelTag": 0,
      "alternatives": [
        {
          "confidence": 0.98267895,
          "transcript": "how old is the Brooklyn Bridge",
        }
      ]
    },
    {
      "channelTag": 1,
      "alternatives": [
        {
          "confidence": 0.98267895,
          "transcript": "where are my glasses",
        }
      ]
    }
  ]
}
```

I think that that's needlessly cumbersome for the user, since the extension doesn't allow for alternatives anyways; I think the user would prefer something like this:
```json
{
  0: ["how old is the Brooklyn Bridge"],
  1: ["where are my glasses"],
}
```

But we can't create a simpler output if we're relying on the API to output the results. So the decision is, basically, what's more worth it? The reliability of the upload, or the simplicity of the output?

Reply:
Having said that, it might be worth it to just drop support for multiple channels, since the most common use case probably won't involve them. That discussion's a bit out of scope, though; let me address all your review comments and see if we can merge this to work on tests.

Reply:
Resolving, we can have further discussion in a subsequent PR.